### ISSUE TITLE
Windows: VST addon does not work if CompCert is also installed

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#9175]
The VST addon copies certain files from CompCert. If both VST and CompCert are installed, VST picks up the wrong version of the files (the one from CompCert), which leads to errors.

I discussed various possible fixes with Andrew and his preference is to ensure via path settings that VST always loads the VST version of the CompCert files. One would have to recompile VST (regenerate .vo files) if one wants to change this but one can assume that people who are picky about which version of CompCert files VST uses anyway compile VST from sources. The bundled addon is more intended for teaching and to get started with VST, where this doesn't matter (in case the CompCert files VST uses are not identical to those CompCert is using, the proof chain breaks at the interface between VST and CompCert).

We also discussed to put the common files between VST and CompCert into a separate hidden installer module, but VST does changes to CompCert files from time to time and this might then lead to release scheduling issues in case the upstream push of the changes to CompCert is not yet done or taged.

**Comment by ejgallego** [comment]
cc: #8052

**Comment by MSoegtropIMC** [comment]
@ejgallego : can you elaborate a bit why #8052 is related? This is not a build issue, but an installer / usability issue.

**Comment by ejgallego** [comment]
Dune will provide a reasonable clean solution to this kind of vensoring problems.

**Comment by MSoegtropIMC** [comment]
@ejgallego : I guess this needs more explanation. The issue is that VST copies certain files from CompCert and can either use its own copies (which might or might not be modified to enable new VST features) or the original CompCert files (possibly reducing the functionality of VST).

Now the Windows installer installs precompiled .vo files. These .vo files are either based on the VST copy of the files or the original copy of the files, and this obviously can't be changed after compiling. So the options are more or less:

1.) always use the VST version of the files (even if CompCert is installed as well)
2.) always use the CompCert version of the files (which means they must be installed with VST in case CompCert is not installed and possibly that lastest VST features cannot be used)
3.) provide two versions of the .vo files and let the user choose somehow which to use, either during installation or by modifying some settings file later.

Andrew want's to go with option 1.

This is mostly an installer and path settings (after install) problem and I don't see how dune would help here. Handling the path settings during compilation is not difficult.

**Comment by ejgallego** [comment]
Thanks @MSoegtropIMC , indeed I think I had understood the problematic here, and Dune is precisely designed to enable option 1 pretty seamlessly, and it can fallback to a system-installed compcert usually in a reasonable way.

**Comment by MSoegtropIMC** [comment]
@ejgallego : sorry, I still don't get it. How could I transport the wisdom of dune through the NSIS installer? How can dune magically change precompiled .vo files so that they are based on different Coq sources so that I won't get "X makes different assumptions about Y" errors when requiring a module which is based on different sources? Unlike Opam, the Windows installer works such that everything (including .vo files) is precompiled and fixed. I really don't see how Dune could help to "fallback to a system-installed compcert" when .vo files are precompiled.

**Comment by Zimmi48** [comment]
Right now, dune won't be of any help. In the future (but not just yet), VST should use dune to build and at this point, it will become easier to handle that kind of things (a dependency that is shipped with the sources). But there is no "wisdom of dune" to transport to the NSIS installer. Dune will just be a build dependency (instead of make) on Windows too.

**Comment by MSoegtropIMC** [comment]
@Zimmi48 : right now for Windows on the Coq user side nothing is build, so make is not used and cannot be replaced with dune. Everything is build on the CI build server and packaged in the installer.

It is definitely thinkable to add support for building .vo files for addons on demand on the user side and in the past the main issue with this was that some build tools require a bash and a fully functional bash requires quite a bit of infrastructure on Windows and at least for teaching and Coq beginners this is not worth the effort. Would dune solve this issue and work without a bash on Windows?

**Comment by Zimmi48** [comment]
>Would dune solve this issue and work without a bash on Windows?

Maybe someday. There has been investigation around that I think. But that wasn't what I was saying. I was talking about simplifying the build steps on CI only. Dune also generates `.install` files which could then be re-used by the Windows installer.

**Comment by ejgallego** [comment]
Dune won't require a shell to build `.vo` libraries or plugins; to build Coq itself it does require `bash`, but if critical we could replace the few scripts that make use of it.

**Comment by MSoegtropIMC** [comment]
It would be of **great** advantage for Windows to have a reliable build system which does not require a bash.

Btw.: I did work on a posix compliant "system command" interpreter. It should be posix compliant for setting environment variables, redirection (even in complicated cases), piping, chaining, ... but it doesn't support high level things like if, for, function. I intended to replace the OCaml system command with this on Windows, so that one can have a system command which does the same on Windows and Linux. I can finish this quickly if it is of any help for the Dune project. I started this to enable native compute in Coq on Windows, which as is essentially also requires a bash.
